timely/src/dataflow/operators/to_stream.rs

```diff
@@ -56,3 +60,108 @@ impl<T: Timestamp, I: IntoIterator+'static> ToStream<T, I::Item> for I where I::
         })
     }
 }
+
+/// Data and progress events of the native stream.
+pub enum Event<F: IntoIterator, D> {
+    /// Indicates that timestamps have advanced to frontier F
+    Progress(F),
+    /// Indicates that event D happened at time T
+    Message(F::Item, D),
+}
+
+/// Converts to a timely `Stream`.
+pub trait ToStreamAsync<T: Timestamp, D: Data> {
+    /// Converts a [native `Stream`](futures_util::stream::Stream) of [`Event`s](Event) into a [timely
+    /// `Stream`](crate::dataflow::Stream). 
+    ///
+    /// # Examples
+    ///
+    /// ```notest,rust
+    /// use timely::dataflow::operators::{Event, ToStreamAsync};
+    /// use timely::dataflow::operators::Inspect;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let iter = (0..4)
+    ///         .map(|x| Event::Message(0, x))
+    ///         .chain(Some(Event::Progress(1)));
+    ///     let stream = futures_util::stream::iter(iter);
+    ///
+    ///     timely::example(move |scope| {
+    ///         stream
+    ///             .to_stream(scope);
+    ///             .inspect(|x| println!("seen: {:?}", x));
+    ///     });
+    /// }
+    /// ```
+    fn to_stream<S: Scope<Timestamp = T>>(self: Pin<Box<Self>>, scope: &S) -> Stream<S, D>;
+}
+
+impl<T, D, F, I> ToStreamAsync<T, D> for I
+where
+    D: Data,
+    T: Timestamp,
+    F: IntoIterator<Item = T>,
+    I: futures_util::stream::Stream<Item = Event<F, D>> + ?Sized + 'static,
+{
+    fn to_stream<S: Scope<Timestamp = T>>(mut self: Pin<Box<Self>>, scope: &S) -> Stream<S, D> {
+        source(scope, "ToStreamAsync", move |capability, info| {
+            let activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+
+            let mut cap_set = CapabilitySet::from_elem(capability);
+
+            move |output| {
+                let waker = futures_util::task::waker_ref(&activator);
+                let mut context = Context::from_waker(&waker);
+
+                // Consume all the ready items of the source_stream and issue them to the operator
+                while let Poll::Ready(item) = self.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Event::Progress(time)) => {
+                            cap_set.downgrade(time);
+                        }
+                        Some(Event::Message(time, data)) => {
+                            output.session(&cap_set.delayed(&time)).give(data);
+                        }
+                        None => {
+                            cap_set.downgrade(&[]);
+                            break;
+                        }
```

Comment:
This is interesting! I suppose it is a reasonable behavior to error-proof use of the stream. It duplicates the information that `Event::Progress([])` represents, but I can't think what else it would mean to end a stream. Elsewhere `None` just means "nothing else at the moment", but since you can presume that the stream won't ever pause like that, I guess we get more information to return.